### PR TITLE
feat: Add clickable links in cell overflow modal

### LINKED ISSUE #1128: Add clickable links in cell overflow modal
As a user, I want links in the cell overflow modal to be styled properly (#712 ) and open the link upon left click. 

Since the editor is using monaco editor, we have to figure out how that can be done.

image

### PR DESCRIPTION
**Pull request by emilyhuxng**
Closes #1128 

Links without `https` are also recognized as links. I couldn't figure out how to disable the tooltip or open the link upon left click.

### CODE REVIEW

**Comment by emilyhuxng** [comment]
> I noticed some inconsistency between clicking e-mail addresses vs. URIs - e-mails try to open in the same window, whereas URIs open in a separate window.

This might be a browser setting? For me, it's doing the same thing as the cells. Urls open in a new tab if you click follow link, redirecting you to the tab. If you control click, it opens in a new tab in the background in the same window. Emails open the mail app.

> I noticed that with links: true (which should be the [default according to docs](https://microsoft.github.io/monaco-editor/typedoc/interfaces/editor.IEditorOptions.html#links)? Not sure why it's not defaulting to clickable) it handles hyperlinking URIs correctly, but doesn't seem to pick up e-mails. Were there any other cases you noticed?

Other than emails, it doesn't detect links without `https` like `google.com`. That is fine, but it would be inconsistent with the cells.

File: packages/console/src/notebook/Editor.tsx [hunk 1]
```diff
@@ -112,6 +114,39 @@ class Editor extends Component<EditorProps, Record<string, never>> {
     this.editor.layout();
     MonacoUtils.removeConflictingKeybindings(this.editor);
 
+    monaco.languages.registerLinkProvider('plaintext', {
+      provideLinks: (model: monaco.editor.ITextModel) => {
+        const text = model.getValue();
```
[the remark this answers is not in the document]
Comment (emilyhuxng) [reply packages/console/src/notebook/Editor.tsx:119]: What about links that span over one line?

**Comment by dsmmcken** [comment]
Ctrl+click is a non-continuous selection behavior in grid, and we wouldn't want to change that.

**Comment by mofojed** [comment]
> > I noticed some inconsistency between clicking e-mail addresses vs. URIs - e-mails try to open in the same window, whereas URIs open in a separate window.
> 
> This might be a browser setting? For me, it's doing the same thing as the cells. Urls open in a new tab if you click follow link, redirecting you to the tab. If you control click, it opens in a new tab in the background in the same window. Emails open the mail app.
Yes, this is true. In Chrome settings in chrome://settings/handlers?search=protocol+handler, I have GMail set up to be the default email handler:
image
Full instructions here: https://www.indeed.com/career-advice/career-development/how-to-get-email-links-to-open-in-chrome
Though now that it's set up, I don't know how to modify that setting... annoying that it tries to open in the same window, but okay.

File: packages/console/src/notebook/Editor.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/console/src/notebook/Editor.tsx:119]: In this case we're only looking for links, which can't span more than one line.

File: packages/console/src/monaco/MonacoUtils.ts [hunk 2]
```diff
@@ -450,6 +451,26 @@ class MonacoUtils {
       KeyCodeUtils.fromString(keyValue)
     );
   }
+
+  static getProvideLinks = () => (model: monaco.editor.ITextModel) => {
```
Comment (mofojed) [inline packages/console/src/monaco/MonacoUtils.ts:455]: No need to curry the method, just have it as a regular method:
```suggestion
  static provideLinks(model: monaco.editor.ITextModel) {
```

File: packages/console/src/notebook/Editor.tsx [hunk 3]
```diff
@@ -112,6 +113,10 @@ class Editor extends Component<EditorProps, Record<string, never>> {
     this.editor.layout();
     MonacoUtils.removeConflictingKeybindings(this.editor);
 
+    monaco.languages.registerLinkProvider('plaintext', {
+      provideLinks: MonacoUtils.getProvideLinks(),
```
Comment (mofojed) [inline packages/console/src/notebook/Editor.tsx:117]: Without the method curried, can just do:
```suggestion
      provideLinks: MonacoUtils.provideLinks,
```

**Review by mofojed — CHANGES_REQUESTED** [review]

**Review by mofojed — CHANGES_REQUESTED** [review]

**Review by dsmmcken — CHANGES_REQUESTED** [review]
Can you check styling to format the links to have $primary color. Not sure if it might be part of the monaco theme variables

**Review by dsmmcken — DISMISSED** [review]

**Review by mofojed — CHANGES_REQUESTED** [review]

File: packages/console/src/notebook/Editor.scss [hunk 4]
```diff
@@ -0,0 +1,5 @@
+@import '@deephaven/components/scss/custom.scss';
+
+.monaco-editor .detected-link {
```
[the remark this answers is not in the document]
Comment (emilyhuxng) [reply packages/console/src/notebook/Editor.scss:3]: I think `textLink.foreground` controls the color of `Follow Link`, `textLink.activeForeground` controls when you hover over `Follow Link` and `editorLink.activeForeground` is when you hold your control key down over a link. I couldn't find anything for the actual visible links.

[the remark this answers is not in the document]
Comment (mofojed) [reply packages/console/src/notebook/Editor.scss:3]: @dsmmcken what styling should we use? For regular links we add in [bootstrap_overrides.scss](https://github.com/deephaven/web-client-ui/blob/e314be6d32792aea3791ee5189fd45d37c86011c/packages/components/scss/bootstrap_overrides.scss#L196):
```
//links
$link-color: $gray-400;
$link-hover-color: $foreground;
```
But then for grid we use `$primary` and then don't add a hover colour.

Should we do the same for Monaco? (`textLink.foreground` and `.detected-link` as `$link-color`, then `textLnk.activeForeground` and `editorLink.activeForeground` as `$link-hover-color`)? Or use `$primary`

[the remark this answers is not in the document]
Comment (dsmmcken) [reply packages/console/src/notebook/Editor.scss:3]: should probably all use link-color and link-hover-color when hovered. I don't think we can set the hover start in monaco as it looks like the link is split into multiple dom elements but I could be wrong.

We should also probably change link color to primary and primary-hover.

**Review by mofojed — APPROVED** [review]

**Pull request merged**